transport/nio/src/test/java/org/kaazing/gateway/transport/tcp/TcpConnectorIT.java

```diff
@@ -0,0 +1,203 @@
+
+    @Before
+    public void before() throws Exception {
+        acceptor = new NioSocketAcceptor(new Properties());
+        acceptor.setSchedulerProvider(schedulerProvider = new SchedulerProvider());
+        acceptor.setResourceAddressFactory(newResourceAddressFactory());
+    }
+
+    private void bindTo8080(IoHandlerAdapter handler) {
+        ResourceAddressFactory addressFactory = ResourceAddressFactory.newResourceAddressFactory();
+
+        Map<String, Object> acceptOptions = new HashMap<>();
+        
+        final String connectURIString = "tcp://127.0.0.1:8080";
+        final ResourceAddress bindAddress =
+                addressFactory.newResourceAddress(
+                        URI.create(connectURIString),
+                        acceptOptions);
+        
+        acceptor.bind(bindAddress, handler, null);
+    }
+
+    @After
+    public void after() throws Exception {
+        // Make sure we always stop all I/O worker threads
+        if (acceptor != null) {
+            schedulerProvider.shutdownNow();
+            acceptor.dispose();
+        }
+    }
+    
+    @Test
+    @Specification({
+        "establish.connection/tcp.client"
+        })
+    public void establishConnection() throws Exception {
+        bindTo8080(new IoHandlerAdapter());
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "server.sent.data/tcp.client"
+        })
+    public void serverSentData() throws Exception {
+        bindTo8080(new IoHandlerAdapter(){
+            @Override
+            protected void doSessionOpened(IoSession session) throws Exception {
+             // KG-8210: push pending write requests onto the write request queue before disposing, should not cause hang in dispose
+                ByteBuffer data = ByteBuffer.allocate(20);
+                String str = "server data";
+                data.put(str.getBytes());
+                
+                data.flip();
+
+                IoBufferAllocatorEx<?> allocator = ((IoSessionEx) session).getBufferAllocator();
+                
+                session.write(allocator.wrap(data.duplicate(), IoBufferEx.FLAG_SHARED));
+            }
+        });
+        
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "client.sent.data/tcp.client"
+        })
+    public void clientSentData() throws Exception {
+        bindTo8080(new IoHandlerAdapter());
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "bidirectional.data/tcp.client"
+        })
+    public void bidirectionalData() throws Exception {
+        bindTo8080(new IoHandlerAdapter(){
+            private int counter = 1;
+            @Override
+            protected void doMessageReceived(IoSession session, Object message) throws Exception {
+             // KG-8210: push pending write requests onto the write request queue before disposing, should not cause hang in dispose
+                ByteBuffer data = ByteBuffer.allocate(20);
```

Comment:
Need to read and verify the incoming messages, accounting for possible fragmentation